In [1]:
import polars as pl

In [2]:
df = pl.read_csv("chembl_mols.csv")
#df.head() #read first 5 rows
df #read full dataset

shape: (2331700, 1)
┌─────────────────────────────────────┐
│ ChEMBL ID";"Name";"Synonyms";"Ty... │
│ ---                                 │
│ str                                 │
╞═════════════════════════════════════╡
│ CHEMBL1206185;";";Small molecule... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL539070;";";Small molecule;... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL3335528;";";Small molecule... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL2419030;";";Small molecule... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ ...                                 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL374652;";";Small molecule;... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1416264;";";Small molecule... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL213734;";";Small molecule;... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1531634;";";Small molecule... │
└─────────────────────────────────────┘

The dataset shows a fully integrated dataframe with all data packed as strings for each compound in each row. Each variable is separated by semicolons. To read the dataframe properly, let's separate them and transform the dataframe to a more readable format.

In [3]:
# Going back to polars documentation, use "sep" to set the delimiter of the file
# which in this case is semicolon
df = pl.read_csv("chembl_mols.csv", sep = ";")
# Show the first 10 rows of data
df.head(10)

shape: (10, 32)
┌────────────┬──────┬──────────┬───────────┬─────┬────────────┬────────────┬───────────┬───────────┐
│ ChEMBL ID  ┆ Name ┆ Synonyms ┆ Type      ┆ ... ┆ Molecular  ┆ Molecular  ┆ Smiles    ┆ Inchi Key │
│ ---        ┆ ---  ┆ ---      ┆ ---       ┆     ┆ Species    ┆ Formula    ┆ ---       ┆ ---       │
│ str        ┆ str  ┆ str      ┆ str       ┆     ┆ ---        ┆ ---        ┆ str       ┆ str       │
│            ┆      ┆          ┆           ┆     ┆ str        ┆ str        ┆           ┆           │
╞════════════╪══════╪══════════╪═══════════╪═════╪════════════╪════════════╪═══════════╪═══════════╡
│ CHEMBL1206 ┆      ┆          ┆ Small     ┆ ... ┆ ACID       ┆ C35H45NO4S ┆ CCCCCCCCC ┆ UFBLKYIDZ │
│ 185        ┆      ┆          ┆ molecule  ┆     ┆            ┆ 2          ┆ CC#CC(N)c ┆ FRLPR-UHF │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ 1ccccc1-c ┆ FFAOYSA-N │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ 1ccc(...  ┆           │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL5390 ┆      ┆          ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C11H15ClN4 ┆ CCCOc1ccc ┆ WPEWNRKLK │
│ 70         ┆      ┆          ┆ molecule  ┆     ┆            ┆ OS         ┆ cc1-c1nnc ┆ LNLSO-UHF │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ (NN)s1.Cl ┆ FFAOYSA-N │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL3335 ┆      ┆          ┆ Small     ┆ ... ┆ ACID       ┆ C41H46O19  ┆ COC(=O)[C ┆ KGUJQZWYZ │
│ 528        ┆      ┆          ┆ molecule  ┆     ┆            ┆            ┆ @H](O[C@@ ┆ PYYRZ-LWE │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ H]1O[C@@H ┆ WUKDVSA-N │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ ](C)[...  ┆           │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL2419 ┆      ┆          ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C14H12F3N3 ┆ O=c1nc(NC ┆ QGDMYSDFC │
│ 030        ┆      ┆          ┆ molecule  ┆     ┆            ┆ O3S        ┆ 2CCCC2)sc ┆ XOKML-UHF │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ 2c([N+](= ┆ FFAOYSA-N │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ O)[O-...  ┆           │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ ...        ┆ ...  ┆ ...      ┆ ...       ┆ ... ┆ ...        ┆ ...        ┆ ...       ┆ ...       │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1969 ┆      ┆          ┆ Small     ┆ ... ┆            ┆            ┆           ┆           │
│ 944        ┆      ┆          ┆ molecule  ┆     ┆            ┆            ┆           ┆           │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL3465 ┆      ┆          ┆ Small     ┆ ... ┆ BASE       ┆ C18H26FN3O ┆ CC(O)CN1C ┆ FZEVYCHTA │
│ 961        ┆      ┆          ┆ molecule  ┆     ┆            ┆            ┆ CC(CN(C)C ┆ DTXPM-UHF │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ c2cc(C#N) ┆ FFAOYSA-N │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ ccc2F...  ┆           │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┤
│ CHEMBL5874 ┆      ┆          ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C26H21F3N4 ┆ Nc1cccc(C ┆ KZOHKPSNJ │
│ 95         ┆      ┆          ┆ molecule  ┆     ┆            ┆ S          ┆ NCc2ccc(- ┆ BXTRJ-UHF │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ c3ccc(-c4 ┆ FFAOYSA-N │
│            ┆      ┆          ┆           ┆     ┆            ┆            ┆ nc5cc...  ┆           │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌

In [4]:
# List all column names to see what compound properties are stored in chEMBL database
df_col_list = list(df.columns)
df_col_list

['ChEMBL ID',
 'Name',
 'Synonyms',
 'Type',
 'Max Phase',
 'Molecular Weight',
 'Targets',
 'Bioactivities',
 'AlogP',
 'Polar Surface Area',
 'HBA',
 'HBD',
 '#RO5 Violations',
 '#Rotatable Bonds',
 'Passes Ro3',
 'QED Weighted',
 'CX Acidic pKa',
 'CX Basic pKa',
 'CX LogP',
 'CX LogD',
 'Aromatic Rings',
 'Structure Type',
 'Inorganic Flag',
 'Heavy Atoms',
 'HBA (Lipinski)',
 'HBD (Lipinski)',
 '#RO5 Violations (Lipinski)',
 'Molecular Weight (Monoisotopic)',
 'Molecular Species',
 'Molecular Formula',
 'Smiles',
 'Inchi Key']

As shown above, a separate dataframe was saved as df_col_list to show all the column names used, which were equivalent to the physicochemical properties of small molecules curated by ChEMBL database.

Selected definitions of the physicochemical properties of compounds were adapted from ChEMBL_31 schema documentation (available as "Release notes" on website) as shown below:

'Max Phase' = Maximum phase of development reached for the compound (4 = approved). Null where max phase has not yet been assigned.

'AlogP' = Calculated partition coefficient

'HBA' = Number of hydrogen bond acceptors

'HBD' = Number of hydrogen bond donors

'#RO5 Violations' = Number of violations of Lipinski's rule-of-five, using HBA and HBD definitions

'Passes Ro3' = Indicates whether the compound passes the rule-of-three (mw < 300, logP < 3 etc)

'QED Weighted' = Weighted quantitative estimate of drug likeness (as defined by Bickerton et al., Nature Chem 2012)

'CX Acidic pKa' = 	The most acidic pKa calculated using ChemAxon v17.29.0

'CX Basic pKa' = The most basic pKa calculated using ChemAxon v17.29.0
 
'CX LogP' = The calculated octanol/water partition coefficient using ChemAxon v17.29.0
 
'CX LogD' = The calculated octanol/water distribution coefficient at pH7.4 using ChemAxon v17.29.0

'Inchi Key' = the IUPAC international chemical identifier key

In [5]:
# Checking if there were any names recorded in the "Name" column by using groupby and count methods
df.groupby("Name", maintain_order = True).agg(pl.count())

shape: (45961, 2)
┌────────────────────────────┬─────────┐
│ Name                       ┆ count   │
│ ---                        ┆ ---     │
│ str                        ┆ u32     │
╞════════════════════════════╪═════════╡
│                            ┆ 2284161 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CLAUSINE F                 ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CLAUSAMINE                 ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ XEROPHILUSIN X1            ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...                        ┆ ...     │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ELAPHOPILOSIN D            ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 15-ACETOXY-EREMANTHOLIDE B ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ SODIUM DIPICOLINATE        ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CLOFENAMIDE                ┆ 1       │
└────────────────────────────┴─────────┘

In [6]:
# Checking if there were anything recorded for the "Synonyms" column
df.groupby("Synonyms", maintain_order = True).agg(pl.count())

shape: (87865, 2)
┌─────────────────────────────────────┬─────────┐
│ Synonyms                            ┆ count   │
│ ---                                 ┆ ---     │
│ str                                 ┆ u32     │
╞═════════════════════════════════════╪═════════╡
│                                     ┆ 2243571 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 25-Deacetyl-Rifabutinol             ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ NSC-648605                          ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ Clausine F                          ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...                                 ┆ ...     │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 4-Tert-Butylflavonol                ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 4-(((2-Aminoethyl)Sulfanyl)(Diph... ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ LUF-5980                            ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 4-CHLORO-M-BENZENEDISULFONAMIDE|... ┆ 1       │
└─────────────────────────────────────┴─────────┘

In [7]:
# Interested in what types of "Max Phase" were recorded for the curated small molecules in ChEMBL database
df.groupby("Max Phase", maintain_order = True).agg(pl.count())

shape: (5, 2)
┌───────────┬─────────┐
│ Max Phase ┆ count   │
│ ---       ┆ ---     │
│ i64       ┆ u32     │
╞═══════════╪═════════╡
│ 0         ┆ 2322126 │
├╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 3         ┆ 1628    │
├╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 4         ┆ 4121    │
├╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 1         ┆ 1479    │
├╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 2         ┆ 2346    │
└───────────┴─────────┘

A quick groupby function showed that there were only 4,121 small molecules approved. Phase 3 recorded a total of 1,628 small molecules. The number was higher for phase 2, which had 2,346 small molecules, followed by 1,479 compounds in phase 1. There were, however, a massive total amount of 2,322,126 small molecules that had zero as phase number, which likely were either waiting to be assigned appropriate phase numbers or the compounds did not enter phase 1 at all.

Next the "Target" column was explored, all entries were shown as numbers. So what did all these numbers represent? I went through several places on ChEMBL database website. Eventually I've settled with an answer that was previously asked by someone, which was relevant or similar to what I wanted to ask (screenshots for ChEMBL cpd related targets). Each compound in the dataset had the number of different targets related to it shown in the "Target" column.

In [8]:
# Here I tried to find out how many times each number of targets appeared in the dataset by
# using the simple groupby and count functions, keeping the dataframe in the original order
df.groupby("Targets", maintain_order = True).agg(pl.count())

shape: (580, 2)
┌─────────┬────────┐
│ Targets ┆ count  │
│ ---     ┆ ---    │
│ str     ┆ u32    │
╞═════════╪════════╡
│         ┆ 96223  │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ 1       ┆ 765614 │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ 2       ┆ 344812 │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ 4       ┆ 169799 │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ ...     ┆ ...    │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ 461     ┆ 1      │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ 395     ┆ 1      │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ 442     ┆ 1      │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┤
│ 474     ┆ 1      │
└─────────┴────────┘

In [9]:
# Sorting the "Targets" column in a descending order
df.sort("Targets", reverse = True)

shape: (2331700, 32)
┌─────────┬───────────┬───────────┬───────────┬─────┬────────────┬────────────┬──────────┬─────────┐
│ ChEMBL  ┆ Name      ┆ Synonyms  ┆ Type      ┆ ... ┆ Molecular  ┆ Molecular  ┆ Smiles   ┆ Inchi   │
│ ID      ┆ ---       ┆ ---       ┆ ---       ┆     ┆ Species    ┆ Formula    ┆ ---      ┆ Key     │
│ ---     ┆ str       ┆ str       ┆ str       ┆     ┆ ---        ┆ ---        ┆ str      ┆ ---     │
│ str     ┆           ┆           ┆           ┆     ┆ str        ┆ str        ┆          ┆ str     │
╞═════════╪═══════════╪═══════════╪═══════════╪═════╪════════════╪════════════╪══════════╪═════════╡
│ CHEMBL1 ┆ DASATINIB ┆ BMS-35482 ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C22H26ClN7 ┆ Cc1nc(Nc ┆ ZBNZXTG │
│ 421     ┆           ┆ 5|BMS-354 ┆ molecule  ┆     ┆            ┆ O2S        ┆ 2ncc(C(= ┆ UTAYRHI │
│         ┆           ┆ 825 HYDRA ┆           ┆     ┆            ┆            ┆ O)Nc3c(C ┆ -UHFFFA │
│         ┆           ┆ TE|BM...  ┆           ┆     ┆            ┆            ┆ )cccc3Cl ┆ OYSA-N  │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ...      ┆         │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1 ┆           ┆           ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C24H15Cl2N ┆ CC1=NN(c ┆ ALZXWBK │
│ 997349  ┆           ┆           ┆ molecule  ┆     ┆            ┆ 5O2        ┆ 2ccc(Cl) ┆ JJOMNHK │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ c(Cl)c2) ┆ -UHFFFA │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ C(=O)C1= ┆ OYSA-N  │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ...      ┆         │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL6 ┆ (+)-TAXIF ┆ (+)-TAXIF ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C15H12O7   ┆ O=C1c2c( ┆ CXQWRCV │
│ 6       ┆ OLIN      ┆ OLIN      ┆ molecule  ┆     ┆            ┆            ┆ O)cc(O)c ┆ TCMQVQX │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ c2O[C@H] ┆ -LSDHHA │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ (c2ccc(O ┆ IUSA-N  │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ...      ┆         │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL5 ┆           ┆           ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C15H20ClNO ┆ Cl.O=C(/ ┆ GWNLEXK │
│ 42969   ┆           ┆           ┆ molecule  ┆     ┆            ┆ 2          ┆ C=C/c1cc ┆ NAOCIOV │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ccc1)CCN ┆ -UHDJGP │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ 1CCOCC1  ┆ CESA-N  │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...     ┆ ...       ┆ ...       ┆ ...       ┆ ... ┆ ...        ┆ ...        ┆ ...      ┆ ...     │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1 ┆           ┆           ┆ Small     ┆ ... ┆ ZWITTERION ┆ C5H13N2O3P ┆ N=C(S)CC ┆ NRCOYHJ │
│ 204556  ┆           ┆           ┆ molecule  ┆     ┆            ┆ S2         ┆ NCCSP(=O ┆ DRQUVCY │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ )(O)O    ┆ -UHFFFA │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆          ┆ OYSA-N  │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL2 ┆           ┆           ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C22H19N3O2 ┆ O=C(CCc1 ┆ XGCNTQK │
│ 365650  ┆           ┆           ┆ molecule  ┆     ┆            ┆            ┆ c[nH]c2c ┆ GPDFKJT │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ cccc12)N ┆ -UHFFFA │
│         ┆           ┆           ┆     

In [10]:
# Then I tried to sort it similarly using "Targets" column with also the "ChEMBL ID" column, 
# showing the highest number of targets recorded appeared to be 994!
df.sort(
    [pl.col("Targets"), pl.col("ChEMBL ID") ** 2],
    reverse = [True, False],
)

shape: (2331700, 32)
┌─────────┬───────────┬───────────┬───────────┬─────┬────────────┬────────────┬──────────┬─────────┐
│ ChEMBL  ┆ Name      ┆ Synonyms  ┆ Type      ┆ ... ┆ Molecular  ┆ Molecular  ┆ Smiles   ┆ Inchi   │
│ ID      ┆ ---       ┆ ---       ┆ ---       ┆     ┆ Species    ┆ Formula    ┆ ---      ┆ Key     │
│ ---     ┆ str       ┆ str       ┆ str       ┆     ┆ ---        ┆ ---        ┆ str      ┆ ---     │
│ str     ┆           ┆           ┆           ┆     ┆ str        ┆ str        ┆          ┆ str     │
╞═════════╪═══════════╪═══════════╪═══════════╪═════╪════════════╪════════════╪══════════╪═════════╡
│ CHEMBL1 ┆ DASATINIB ┆ BMS-35482 ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C22H26ClN7 ┆ Cc1nc(Nc ┆ ZBNZXTG │
│ 421     ┆           ┆ 5|BMS-354 ┆ molecule  ┆     ┆            ┆ O2S        ┆ 2ncc(C(= ┆ UTAYRHI │
│         ┆           ┆ 825 HYDRA ┆           ┆     ┆            ┆            ┆ O)Nc3c(C ┆ -UHFFFA │
│         ┆           ┆ TE|BM...  ┆           ┆     ┆            ┆            ┆ )cccc3Cl ┆ OYSA-N  │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ...      ┆         │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1 ┆           ┆           ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C24H15Cl2N ┆ CC1=NN(c ┆ ALZXWBK │
│ 997349  ┆           ┆           ┆ molecule  ┆     ┆            ┆ 5O2        ┆ 2ccc(Cl) ┆ JJOMNHK │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ c(Cl)c2) ┆ -UHFFFA │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ C(=O)C1= ┆ OYSA-N  │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ...      ┆         │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL6 ┆ (+)-TAXIF ┆ (+)-TAXIF ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C15H12O7   ┆ O=C1c2c( ┆ CXQWRCV │
│ 6       ┆ OLIN      ┆ OLIN      ┆ molecule  ┆     ┆            ┆            ┆ O)cc(O)c ┆ TCMQVQX │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ c2O[C@H] ┆ -LSDHHA │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ (c2ccc(O ┆ IUSA-N  │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ...      ┆         │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL5 ┆           ┆           ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C15H20ClNO ┆ Cl.O=C(/ ┆ GWNLEXK │
│ 42969   ┆           ┆           ┆ molecule  ┆     ┆            ┆ 2          ┆ C=C/c1cc ┆ NAOCIOV │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ccc1)CCN ┆ -UHDJGP │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ 1CCOCC1  ┆ CESA-N  │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...     ┆ ...       ┆ ...       ┆ ...       ┆ ... ┆ ...        ┆ ...        ┆ ...      ┆ ...     │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1 ┆           ┆           ┆ Small     ┆ ... ┆ ZWITTERION ┆ C5H13N2O3P ┆ N=C(S)CC ┆ NRCOYHJ │
│ 204556  ┆           ┆           ┆ molecule  ┆     ┆            ┆ S2         ┆ NCCSP(=O ┆ DRQUVCY │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ )(O)O    ┆ -UHFFFA │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆          ┆ OYSA-N  │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL2 ┆           ┆           ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C22H19N3O2 ┆ O=C(CCc1 ┆ XGCNTQK │
│ 365650  ┆           ┆           ┆ molecule  ┆     ┆            ┆            ┆ c[nH]c2c ┆ GPDFKJT │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ cccc12)N ┆ -UHFFFA │
│         ┆           ┆           ┆     

In [11]:
# This was confirmed via a filter function, which brought up CHEMBL1421, or also known as dasatinib
df.filter(pl.col("Targets") == "994")

shape: (1, 32)
┌─────────┬───────────┬───────────┬───────────┬─────┬────────────┬────────────┬──────────┬─────────┐
│ ChEMBL  ┆ Name      ┆ Synonyms  ┆ Type      ┆ ... ┆ Molecular  ┆ Molecular  ┆ Smiles   ┆ Inchi   │
│ ID      ┆ ---       ┆ ---       ┆ ---       ┆     ┆ Species    ┆ Formula    ┆ ---      ┆ Key     │
│ ---     ┆ str       ┆ str       ┆ str       ┆     ┆ ---        ┆ ---        ┆ str      ┆ ---     │
│ str     ┆           ┆           ┆           ┆     ┆ str        ┆ str        ┆          ┆ str     │
╞═════════╪═══════════╪═══════════╪═══════════╪═════╪════════════╪════════════╪══════════╪═════════╡
│ CHEMBL1 ┆ DASATINIB ┆ BMS-35482 ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C22H26ClN7 ┆ Cc1nc(Nc ┆ ZBNZXTG │
│ 421     ┆           ┆ 5|BMS-354 ┆ molecule  ┆     ┆            ┆ O2S        ┆ 2ncc(C(= ┆ UTAYRHI │
│         ┆           ┆ 825 HYDRA ┆           ┆     ┆            ┆            ┆ O)Nc3c(C ┆ -UHFFFA │
│         ┆           ┆ TE|BM...  ┆           ┆     ┆            ┆            ┆ )cccc3Cl ┆ OYSA-N  │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ...      ┆         │
└─────────┴───────────┴───────────┴───────────┴─────┴────────────┴────────────┴──────────┴─────────┘

In [12]:
# Then I tested the filter function, so I searched for compounds with one target only, 
# zero violations of the rule-of-five and achieved a max phase of 4 which meant they were approved compounds
# This brought up a total of 92 compounds as shown below
df.filter((pl.col("Targets") == "1") & (pl.col("#RO5 Violations") == "0") & (pl.col("Max Phase") == 4))

shape: (92, 32)
┌─────────┬───────────┬───────────┬───────────┬─────┬────────────┬────────────┬──────────┬─────────┐
│ ChEMBL  ┆ Name      ┆ Synonyms  ┆ Type      ┆ ... ┆ Molecular  ┆ Molecular  ┆ Smiles   ┆ Inchi   │
│ ID      ┆ ---       ┆ ---       ┆ ---       ┆     ┆ Species    ┆ Formula    ┆ ---      ┆ Key     │
│ ---     ┆ str       ┆ str       ┆ str       ┆     ┆ ---        ┆ ---        ┆ str      ┆ ---     │
│ str     ┆           ┆           ┆           ┆     ┆ str        ┆ str        ┆          ┆ str     │
╞═════════╪═══════════╪═══════════╪═══════════╪═════╪════════════╪════════════╪══════════╪═════════╡
│ CHEMBL1 ┆ PHENMETRA ┆ NSC-40572 ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C11H16ClNO ┆ CC1NCCOC ┆ VJNXVAV │
│ 200483  ┆ ZINE HYDR ┆ 8|PHENMET ┆ molecule  ┆     ┆            ┆            ┆ 1c1ccccc ┆ KCZJOFQ │
│         ┆ OCHLORIDE ┆ RALINE    ┆           ┆     ┆            ┆            ┆ 1.Cl     ┆ -UHFFFA │
│         ┆           ┆ HYDROCH.. ┆           ┆     ┆            ┆            ┆          ┆ OYSA-N  │
│         ┆           ┆ .         ┆           ┆     ┆            ┆            ┆          ┆         │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1 ┆ HYDROCORT ┆ HYDROCORT ┆ Small     ┆ ... ┆ NEUTRAL    ┆ C27H41NO6  ┆ CCN(CC)C ┆ FWFVLWG │
│ 201263  ┆ AMATE     ┆ AMATE|Hyd ┆ molecule  ┆     ┆            ┆            ┆ C(=O)OCC ┆ EFDIZMJ │
│         ┆           ┆ rocortama ┆           ┆     ┆            ┆            ┆ (=O)[C@@ ┆ -FOMYWI │
│         ┆           ┆ te        ┆           ┆     ┆            ┆            ┆ ]1(O)CC[ ┆ RZSA-N  │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ...      ┆         │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1 ┆ FOSPROPOF ┆ FOSPROPOF ┆ Small     ┆ ... ┆ ACID       ┆ C13H21O5P  ┆ CC(C)c1c ┆ QVNNONO │
│ 201766  ┆ OL        ┆ OL        ┆ molecule  ┆     ┆            ┆            ┆ ccc(C(C) ┆ FASOXQV │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ C)c1OCOP ┆ -UHFFFA │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ (=O)(O)O ┆ OYSA-N  │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1 ┆ IODINE    ┆ IODINE|IO ┆ Small     ┆ ... ┆ None       ┆ I2         ┆ II       ┆ PNDPGZB │
│ 201225  ┆           ┆ DINE MILD ┆ molecule  ┆     ┆            ┆            ┆          ┆ MCMUPRI │
│         ┆           ┆ |IODINE   ┆           ┆     ┆            ┆            ┆          ┆ -UHFFFA │
│         ┆           ┆ STRONG... ┆           ┆     ┆            ┆            ┆          ┆ OYSA-N  │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...     ┆ ...       ┆ ...       ┆ ...       ┆ ... ┆ ...        ┆ ...        ┆ ...      ┆ ...     │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1 ┆ CARBOPROS ┆ CARBOPROS ┆ Small     ┆ ... ┆ ACID       ┆ C25H47NO8  ┆ CCCCC[C@ ┆ UMMADZJ │
│ 237105  ┆ T TROMETH ┆ T TROMETA ┆ molecule  ┆     ┆            ┆            ┆ ](C)(O)/ ┆ LZAPZAW │
│         ┆ AMINE     ┆ MOL|CARBO ┆           ┆     ┆            ┆            ┆ C=C/[C@H ┆ -XOWPVR │
│         ┆           ┆ PROST...  ┆           ┆     ┆            ┆            ┆ ]1[C@H]( ┆ JPSA-N  │
│         ┆           ┆           ┆           ┆     ┆            ┆            ┆ ...      ┆         │
├╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ CHEMBL1 ┆ VILANTERO ┆ GW-642444 ┆ Small     ┆ ... ┆ BASE       ┆ C24H33Cl2N ┆ OCc1cc([ ┆ DAFYYTQ │
│ 198857  ┆ L         ┆ |GW-64244 ┆ molecule  ┆     ┆            ┆ O5         ┆ C@@H](O) ┆ WSAWIGS │
│         ┆           ┆ 4X|GW6424 ┆           ┆     ┆            ┆            ┆ CNCCCCCC ┆ -DEOSSO │
│         ┆           ┆ 44|GW...  ┆          